convert array access into a terms query + generic function

seut opened the pull request:
Lucene itself doesn’t support to filter on an array element, therefore
an array access was resulting in the generic function filter fallback.
By using a terms query in conjunction with a generic function filter,
Lucene can pre-filter which results in a huge performance increase.

mfussenegger commented on sql/src/main/java/io/crate/lucene/LuceneQueryBuilder.java:
```diff
@@ -1126,6 +1127,57 @@ public Query apply(Function parent, Function inner, Context context) {
+                    MappedFieldType fieldType = context.getFieldTypeOrNull(reference.column().fqn());
+                    if (fieldType == null) {
+                        if (CollectionType.unnest(reference.valueType()).equals(DataTypes.OBJECT)) {
+                            return null; // fallback to generic query to enable {x=10} = any(objects)
```
Should we have a test for this case? And the example in the comment seems wrong?

mfussenegger commented on sql/src/main/java/io/crate/lucene/LuceneQueryBuilder.java:
```diff
@@ -1126,6 +1127,57 @@ public Query apply(Function parent, Function inner, Context context) {
 
+        static class SubscriptQuery implements InnerFunctionToQuery {
+
+            private static final ImmutableMap<String, BooleanClause.Occur> FUNCTION_TO_CLAUSE_OCCURRENCE =
```
Without reading the code below it's not visible why `Gt` and `Lt` have `MUST_NOT` but all others have `MUST`. Maybe we can clarify that a bit. Maybe renaming to something like `PRE_FILTER_OCCURRENCE_BY_OP` or `NARROWING_CLAUSE_BY_OP`  would help to indicate that these are for the additional "pre-filter" clauses.

mfussenegger commented:
Do we have some benchmark results already showing how much of a speed-up this can yield? If so - can we include an excerpt in the commit message?

mfussenegger commented:
retest this please

seut commented:
I'll add benchmark results asap.

seut commented:
@mfussenegger fixups pushed, I've also added the benchmark results to the original commit message.

matriv commented:
@seut sorry, Acked here by mistake, didn't review the PR. sorry!

seut commented:
@mfussenegger As discussed, building a term query for non-eq comparators is wrong. I've changed them to range queries and also added integration tests to validate the behavior. Please have a look again.

mfussenegger commented on sql/src/main/java/io/crate/lucene/LuceneQueryBuilder.java:
```diff
@@ -1126,6 +1127,78 @@ public Query apply(Function parent, Function inner, Context context) {
+                    MappedFieldType fieldType = context.getFieldTypeOrNull(reference.column().fqn());
+                    if (fieldType == null) {
+                        if (CollectionType.unnest(reference.valueType()).equals(DataTypes.OBJECT)) {
+                            return null; // fallback to generic query to enable objects[1] =/</<=/>/>= {x=10}
```
Added the `=/</<=/>/>=` accidentally?
